src/scratch/ScratchRuntime.as

```diff
@@ -89,49 +145,342 @@ public class ScratchRuntime {
+	private var position:int;
+	private var secs:Number;
+	private var alreadyDone:int;
+	
+	private var pSound:Boolean;
+	private var mSound:Boolean;
+	private var mouse:Boolean;
+	private var cursor:Boolean;
+	public var fullEditor:Boolean;
+	private var framerate:Number;
+	private var width:int;
+	private var height:int;
+	public var ready:int=-1;
+	
+	private var mBytes:ByteArray;
+	private var mPosition:int = 0;
+	private var mic:Microphone;
+	private var mReady:Boolean;
+	
+	private var timeout:int;
+	
 	private function saveFrame():void {
-		var f:BitmapData = new BitmapData(480, 360);
-		f.draw(app.stagePane);
-		frames.push(f);
-		if ((frames.length % 100) == 0) {
-			trace('frames: ' + frames.length + ' mem: ' + System.totalMemory);
+		saveSound();
+		var t:Number = getTimer()*.001-secs;
+		while (t>sounds.length/framerate+1/framerate) {
+			saveSound();
+		}
+		if (cursor) var cursor:DisplayObject = Resources.createDO('videoCursor');
+		if (mouse && app.gh.mouseIsDown) var circle:Bitmap = Resources.createBmp('mouseCircle');
+		var f:BitmapData;
+		if (fullEditor) {
+			var aWidth:int = app.stage.stageWidth;
+			var aHeight:int = app.stage.stageHeight;
+			if (!Scratch.app.isIn3D) {
+				if (app.stagePane.videoImage) app.stagePane.videoImage.visible = false;
+			}
+			if (width!=aWidth || height!=aHeight) {
+				var scale:Number = 1.0;
+				scale = width/aWidth > height/aHeight ? height/aHeight : width/aWidth;
+				var m:Matrix = new Matrix();
+				m.scale(scale,scale);
+				f = new BitmapData(width,height,false);
+				f.draw(app.stage,m,null, null, new Rectangle(0,0,aWidth*scale,aHeight*scale),false);
+				if(Scratch.app.isIn3D) {
+					var scaled:Number = scale;
+					if (!app.editMode) {
+						scaled *= app.presentationScale;
+					}
+					else if (app.stageIsContracted) {
+						scaled*=0.5;
+					}
+					var d:BitmapData = app.stagePane.saveScreenData();
+					f.draw(d, new Matrix( scaled, 0, 0, scaled, app.stagePane.localToGlobal(new Point(0, 0)).x*scale, app.stagePane.localToGlobal(new Point(0, 0)).y*scale));
+				}
+				else if (app.stagePane.videoImage) app.stagePane.videoImage.visible = true;
+				if (mouse && app.gh.mouseIsDown) {
+					f.draw(circle,new Matrix(scale,0,0,scale,(app.stage.mouseX-circle.width/2.0)*scale,(app.stage.mouseY-circle.height/2.0)*scale));
+				}
+				if (cursor) {
+					f.draw(cursor,new Matrix(scale,0,0,scale,app.stage.mouseX*scale,app.stage.mouseY*scale));
+				}
+			}
+			else {
+				f = new BitmapData(width,height,false);
+				f.draw(app.stage);
+				if(Scratch.app.isIn3D) {
+					var scaler:Number = 1;
+					if (!app.editMode) {
+						scaler *= app.presentationScale;
+					}
+					else if (app.stageIsContracted) {
+						scaler*=0.5;
+					}
+					var e:BitmapData = app.stagePane.saveScreenData();
+					if (scaler==1) f.copyPixels(e, new Rectangle(0,0,486,432),new Point(app.stagePane.localToGlobal(new Point(0, 0)).x, app.stagePane.localToGlobal(new Point(0, 0)).y));
```

Comment:
Why `486,432`? Should those numbers be given named constants, or maybe calculated from other UI elements' locations and sizes?